core/server/src/main/java/alluxio/master/journal/AsyncJournalWriter.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the “License”). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.journal;
+
+import alluxio.Constants;
+import alluxio.master.MasterContext;
+import alluxio.proto.journal.Journal.JournalEntry;
+
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * This enables async journal writing, as well as some batched journal flushing.
+ */
+public final class AsyncJournalWriter {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private final JournalWriter mJournalWriter;
+  private final ConcurrentLinkedQueue<JournalEntry> mQueue;
+  private final AtomicLong mCounter;
+  private final AtomicLong mFlushCounter;
+  private final long mFlushBatchTime;
+
+  /**
+   * Use a {@link ReentrantLock} to guard the journal writing. Using the fairness policy seems to
+   * result in better throughput.
+   */
+  private final ReentrantLock mFlushLock = new ReentrantLock(true);
+
+  /**
+   * Creates a {@link AsyncJournalWriter}.
+   *
+   * @param journalWriter the {@link JournalWriter} to use for writing
+   */
+  public AsyncJournalWriter(JournalWriter journalWriter) {
+    mJournalWriter = Preconditions.checkNotNull(journalWriter);
+    mQueue = new ConcurrentLinkedQueue<>();
+    mCounter = new AtomicLong(0);
+    mFlushCounter = new AtomicLong(0);
+    mFlushBatchTime = MasterContext.getConf().getLong(Constants.MASTER_JOURNAL_FLUSH_BATCH_TIME_MS);
+  }
+
+  /**
+   * Appends a {@link JournalEntry} for writing to the journal.
+   *
+   * @param entry the {@link JournalEntry} to append
+   * @return a counter for the entry, for flushing
+   */
+  public long appendEntry(JournalEntry entry) {
+    mCounter.incrementAndGet();
+    mQueue.offer(entry);
+    return mCounter.get();
```

Comment:
I added some comments to the code, but only using `incrementAndGet` will not be correct. That get will only be correct up to that point in the code, but the entry has not been added yet. Therefore, the value has to be read AFTER the entry is added.

On the flip side, the counter must be incremented BEFORE the entry is added to stay safe.